contrib/dimfeld/httptreemux.v5: fix route and name for 30X redirects

devillecodes opened the pull request:
The httptreemux router has a redirect behaviour that is invoked when a request URL and matched route only differs in a trailing slash. The default behaviour is to respond with a 301 (moved permanently) to redirect the client to the exact path of the matched handler. We previously patched one of the two scenarios where this occurs in #2332. The changes in this commit addresses the other scenario for the 301 redirect, but also the 307 and 308 redirect options. Several tests were included to cover the various scenarios.

~We also standardize the resource name in the event that there is a trailing slash missmatch between request URL and matched handler. We always truncate the trailing slash in the resource name.~ We no longer do this as part of these changes.

Fixes #2663

<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Commit and PR titles should be prefixed with the general area of the pull request's change.

-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

There are two main changes in this PR:

1. Fix of the other scenario that was missed when this issue was first raised as #2332.
2. Standardized the route and resource name by trimming any unnecessary trailing slashes.

Extensive tests have been added for the described changes.

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
* If this resolves a GitHub issue, include "Fixes #XXXX" to link the issue and auto-close it on merge.
-->

Fixes #2663.

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).
-->

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.


Unsure? Have a question? Request a review!

laughingman-hass commented on contrib/dimfeld/httptreemux.v5/httptreemux.go:
```diff
@@ -112,18 +112,26 @@ func getRoute(router *httptreemux.TreeMux, w http.ResponseWriter, req *http.Requ
+		} else {
+			// if the route does not have a trailing slash, add one
+			rReq.RequestURI = rReq.RequestURI + "/"
+			rReq.URL.Path = rReq.URL.Path + "/"
```
Does this mean we also redirect when there is no trailing slash?

devillecodes replied:
Yes. That's correct. This is the case we missed originally.

rarguelloF commented on contrib/dimfeld/httptreemux.v5/httptreemux_test.go:
```diff
@@ -238,18 +238,777 @@ func TestNamingSchema(t *testing.T) {
 	namingschematest.NewHTTPServerTest(genSpans, "http.router")(t)
 }
 
-func TestTrailingSlashRoutes(t *testing.T) {
```
While tests are highly appreciated, it would be preferred to keep this PR as short / simple as possible to make it easier to review and maintain.

Keeping the existing `TestTrailingSlashRoutes` test as it is with a new case covering this bug should be enough.

devillecodes replied:
I can definitely look into reducing some of the redundant test cases, but I do think at least some additional test cases for the 4 router redirect modes is warranted, i.e. (from `dimfeld/httptreemux/v5/router.go`):

```golang
package httptreemux

const (
	Redirect301 RedirectBehavior = iota // Return 301 Moved Permanently
	Redirect307                         // Return 307 HTTP/1.1 Temporary Redirect
	Redirect308                         // Return a 308 RFC7538 Permanent Redirect
	UseHandler                          // Just call the handler function
)
```

devillecodes replied:
The test cases were reduced down to three scenarios per redirect mode:
1. path not found
2. path with parameterized URL _excluding_ a trailing slash
3. path with parameterized URL _including_ a trailing slash

devillecodes commented on contrib/dimfeld/httptreemux.v5/httptreemux.go:
```diff
@@ -139,5 +147,31 @@ func getRoute(router *httptreemux.TreeMux, w http.ResponseWriter, req *http.Requ
 		route = strings.Replace(route, oldP, newP, 1)
 	}
+
+	// remove trailing slash from route to standardize returned value
+	// the router does not allow you to register two matching routes with the only difference being a trailing slash
+	// this only affects the resulting returned value and not the actual request URL set on tag http.url
+	if trailingSlash {
+		route = strings.TrimSuffix(route, "/")
+	}
```
Ah, this is an excellent point. I will remove this and leave the trailing slash as is.

- [ ] do not trim trailing slash from path

devillecodes commented on contrib/dimfeld/httptreemux.v5/httptreemux.go:
```diff
@@ -139,5 +147,31 @@ func getRoute(router *httptreemux.TreeMux, w http.ResponseWriter, req *http.Requ
 	return route, true
 }
+
+// isSupportedRedirectStatus checks if the given HTTP status code is a supported redirect status.
+// It returns true if the status code is either StatusMovedPermanently, StatusTemporaryRedirect, or StatusPermanentRedirect.
+// Otherwise, it returns false.
```
Good point. I got a bit carried away after discovering what was actually happening here. 😅 

- [ ] remove redundant comments

devillecodes commented on contrib/dimfeld/httptreemux.v5/httptreemux.go:
```diff
@@ -139,5 +147,31 @@ func getRoute(router *httptreemux.TreeMux, w http.ResponseWriter, req *http.Requ
+		status == http.StatusPermanentRedirect
+}
+
+// routerRedirectEnabled checks if the redirection is enabled on the router.
+// It returns true if either RedirectCleanPath or RedirectTrailingSlash is enabled,
+// and the RedirectBehavior is not set to UseHandler. Otherwise, it returns false.
+// This function is used to determine whether to perform redirections based on the router's configuration.
```
- [ ] simplify comment

darccio reviewed (approved):
LGTM

devillecodes commented:
@rarguelloF I have addressed the feedback. Please take a look when you have a moment. Also let me know if there are any follow up actions you'd like me to look at.

rarguelloF reviewed (approved):
LGTM!

devillecodes commented:
Hi folks. Happy Monday! Just checking in if there's anything required from me on this PR at this stage?

darccio commented:
@devillecodes No, it just needs to be merged. Allow us to take care of it during this week, before the beginning of `v1.64.0` release cc @nsrip-dd

darccio commented:
@devillecodes Please, can you merge `main` into your PR? I ran the tests locally and in draft PR #2685 (as external contributors' PRs don't run all the CI). Everything is green.

devillecodes commented:
Amazing! Thanks folks! :raised_hands: Appreciate the efforts in getting this merged! :tada:
